packages/pug-code-gen/index.js

```diff
@@ -767,6 +768,37 @@ Compiler.prototype = {
     this.buf.push('  }\n}).call(this);\n');
   },
 
+  visitEachOf: function(each){
+    var valVarName = 'pug_val' + this.eachOfCount;
+    if (each.key) {
+      var keyVarName = 'pug_key' + this.eachOfCount
+    }
+    this.eachOfCount++;
+
+    this.buf.push(''
+      + '// iterate ' + each.obj + '\n'
+      + ';(function(){\n'
+      + '  var $$obj = ' + each.obj + ';\n'
+      + '  var $$l = 0;\n')
+    if (each.key) {
+      this.buf.push('  for (const [' + keyVarName + ', ' + valVarName + '] of $$obj) {\n');
+    } else {
+      this.buf.push('  for (const ' + valVarName + ' of $$obj) {\n');
+    }
+    this.buf.push('    $$l++;\n');
+    if (each.key) {
+      this.buf.push(''
+        + '    var ' + each.key + ' = ' + valVarName + ';'
+        + '    var ' + each.val + ' = ' + keyVarName + ';');
+    } else {
+      this.buf.push('    var ' + each.val + ' = ' + valVarName + ';')
+    }
+
+    this.visit(each.block, each);
+
+    this.buf.push('  }\n}).call(this);\n');
```

Comment:
If you make the changes proposed above, there will be no variables that are not block scoped to the loop anyway, so we can remove the self invoking function that wraps the loop.